---
title: 終止錯誤 |Microsoft Docs
ms.date: 09/13/2016
ms.openlocfilehash: 272e6cdd9a1da3cfd2e4f730f6aeb27577948278
ms.sourcegitcommit: 0907b8c6322d2c7c61b17f8168d53452c8964b41
ms.translationtype: MT
ms.contentlocale: zh-TW
ms.lasthandoff: 08/05/2020
ms.locfileid: "87786385"
---
# <a name="terminating-errors"></a><span data-ttu-id="ae586-102">終止錯誤</span><span class="sxs-lookup"><span data-stu-id="ae586-102">Terminating Errors</span></span>

<span data-ttu-id="ae586-103">本主題討論用來報告終止錯誤的方法。</span><span class="sxs-lookup"><span data-stu-id="ae586-103">This topic discusses the method used to report terminating errors.</span></span> <span data-ttu-id="ae586-104">此外，它也會討論如何從 Cmdlet 內呼叫方法，並討論當呼叫方法時，Windows PowerShell 執行時間可以傳回的例外狀況。</span><span class="sxs-lookup"><span data-stu-id="ae586-104">It also discusses how to call the method from within the cmdlet, and it discusses the exceptions that can be returned by the Windows PowerShell runtime when the method is called.</span></span>

<span data-ttu-id="ae586-105">當發生終止錯誤時，此 Cmdlet 應該會呼叫[Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError)方法來報告錯誤。</span><span class="sxs-lookup"><span data-stu-id="ae586-105">When a terminating error occurs, the cmdlet should report the error by calling the [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) method.</span></span> <span data-ttu-id="ae586-106">這個方法可讓 Cmdlet 傳送描述導致終止錯誤之條件的錯誤記錄。</span><span class="sxs-lookup"><span data-stu-id="ae586-106">This method allows the cmdlet to send an error record that describes the condition that caused the terminating error.</span></span> <span data-ttu-id="ae586-107">如需錯誤記錄的詳細資訊，請參閱[Windows PowerShell 錯誤記錄](./windows-powershell-error-records.md)。</span><span class="sxs-lookup"><span data-stu-id="ae586-107">For more information about error records, see [Windows PowerShell Error Records](./windows-powershell-error-records.md).</span></span>

<span data-ttu-id="ae586-108">當呼叫[Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError)方法時，Windows PowerShell 執行時間會永久停止執行管線，並擲回[Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException)例外狀況（exception）的功能。</span><span class="sxs-lookup"><span data-stu-id="ae586-108">When the [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) method is called, the  Windows PowerShell runtime permanently stops the execution of the pipeline and throws a [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span></span> <span data-ttu-id="ae586-109">任何後續嘗試呼叫[WriteObject](/dotnet/api/System.Management.Automation.Cmdlet.WriteObject)、 [WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError)或數個其他 api 時，都會導致這些呼叫擲回[Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException)例外狀況（exception），而不會發生問題。</span><span class="sxs-lookup"><span data-stu-id="ae586-109">Any subsequent attempts to call [System.Management.Automation.Cmdlet.WriteObject](/dotnet/api/System.Management.Automation.Cmdlet.WriteObject), [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError), or several other APIs causes those calls to throw a [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span></span>

<span data-ttu-id="ae586-110">如果管線中的另一個 Cmdlet 回報終止錯誤、使用者已要求停止管線，或因為任何原因而在完成之前已停止管線，也可能會發生[Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException)例外狀況。</span><span class="sxs-lookup"><span data-stu-id="ae586-110">The [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception can also occur if another cmdlet in the pipeline reports a terminating error, if the user has asked to stop the pipeline, or if the pipeline has been halted before completion for any reason.</span></span> <span data-ttu-id="ae586-111">Cmdlet 不需要攔截[Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException)例外狀況，除非它必須清除開啟的資源或其內部狀態。</span><span class="sxs-lookup"><span data-stu-id="ae586-111">The cmdlet does not need to catch the [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception unless it must clean up open resources or its internal state.</span></span>

<span data-ttu-id="ae586-112">Cmdlet 可以在報告終止錯誤之前，撰寫任何數目的輸出物件或非終止錯誤。</span><span class="sxs-lookup"><span data-stu-id="ae586-112">Cmdlets can write any number of output objects or non-terminating errors before reporting a terminating error.</span></span> <span data-ttu-id="ae586-113">不過，終止錯誤會永久停止管線，而不會報告進一步的輸出、終止錯誤或非終止錯誤。</span><span class="sxs-lookup"><span data-stu-id="ae586-113">However, the terminating error permanently stops the pipeline, and no further output, terminating errors, or non-terminating errors can be reported.</span></span>

<span data-ttu-id="ae586-114">Cmdlet 只能從呼叫 BeginProcessing、ProcessRecord 或[system.web](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing)的執行緒[Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) ，而此輸入處理方法則是由它所呼叫的程式所組成，而這些都是由稱為[System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing)、 [system.web](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)。</span><span class="sxs-lookup"><span data-stu-id="ae586-114">Cmdlets can call [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) only from the thread that called the [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), or [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) input processing method.</span></span> <span data-ttu-id="ae586-115">請不要嘗試從另一個執行緒呼叫[Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError)或[WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError) ，而是從其他執行緒。</span><span class="sxs-lookup"><span data-stu-id="ae586-115">Do not attempt to call [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) or [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError) from another thread.</span></span> <span data-ttu-id="ae586-116">相反地，錯誤必須傳回到主執行緒。</span><span class="sxs-lookup"><span data-stu-id="ae586-116">Instead, errors must be communicated back to the main thread.</span></span>

<span data-ttu-id="ae586-117">Cmdlet 可能會在它的[BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing)、 [ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)或 system.servicemodel 方法的執行中擲回例外狀況（exception），或使用... [（系統管理](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing).）。</span><span class="sxs-lookup"><span data-stu-id="ae586-117">It is possible for a cmdlet to throw an exception in its implementation of the [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), or [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) method.</span></span> <span data-ttu-id="ae586-118">從這些方法擲回的任何例外狀況 (除了一些停止 Windows PowerShell 主機) 的嚴重錯誤情況，也會被視為終止的錯誤，這會停止管線，而不是 Windows PowerShell 整體。</span><span class="sxs-lookup"><span data-stu-id="ae586-118">Any exception thrown from these methods (except for a few severe error conditions that stop the Windows PowerShell host) is interpreted as a terminating error which stops the pipeline, but not Windows PowerShell as a whole.</span></span> <span data-ttu-id="ae586-119"> (這僅適用于主要 Cmdlet 執行緒。</span><span class="sxs-lookup"><span data-stu-id="ae586-119">(This applies only to the main cmdlet thread.</span></span> <span data-ttu-id="ae586-120">在 Cmdlet 衍生的執行緒中未攔截到的例外狀況，通常會暫停 Windows PowerShell 主機。 ) 建議您使用[Throwterminatingerror \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) ，而不是擲回例外狀況，因為錯誤記錄會提供錯誤條件的其他相關資訊，這對使用者很有用。</span><span class="sxs-lookup"><span data-stu-id="ae586-120">Uncaught exceptions in threads spawned by the cmdlet, in general, halt the Windows PowerShell host.) We recommend that you use [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) rather than throwing an exception because the error record provides additional information about the error condition, which is useful to the end-user.</span></span> <span data-ttu-id="ae586-121">Cmdlet 應遵循 managed 程式碼指導方針，以攔截和處理 (`catch (Exception e)`) 的所有例外狀況。</span><span class="sxs-lookup"><span data-stu-id="ae586-121">Cmdlets should honor the managed code guideline against catching and handling all exceptions (`catch (Exception e)`).</span></span> <span data-ttu-id="ae586-122">只將已知和預期類型的例外狀況轉換成錯誤記錄。</span><span class="sxs-lookup"><span data-stu-id="ae586-122">Convert only exceptions of known and expected types into error records.</span></span>

## <a name="see-also"></a><span data-ttu-id="ae586-123">另請參閱</span><span class="sxs-lookup"><span data-stu-id="ae586-123">See Also</span></span>

[<span data-ttu-id="ae586-124">BeginProcessing 的管理元件</span><span class="sxs-lookup"><span data-stu-id="ae586-124">System.Management.Automation.Cmdlet.BeginProcessing</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing)

[<span data-ttu-id="ae586-125">System.servicemodel. Cmdlet。</span><span class="sxs-lookup"><span data-stu-id="ae586-125">System.Management.Automation.Cmdlet.EndProcessing</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing)

[<span data-ttu-id="ae586-126">ProcessRecord 的管理元件</span><span class="sxs-lookup"><span data-stu-id="ae586-126">System.Management.Automation.Cmdlet.ProcessRecord</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)

[<span data-ttu-id="ae586-127">Pipelinestoppedexception。</span><span class="sxs-lookup"><span data-stu-id="ae586-127">System.Management.Automation.Pipelinestoppedexception</span></span>](/dotnet/api/System.Management.Automation.PipelineStoppedException)

[<span data-ttu-id="ae586-128">Throwterminatingerror \* 的 \*</span><span class="sxs-lookup"><span data-stu-id="ae586-128">System.Management.Automation.Cmdlet.Throwterminatingerror\*</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError)

[<span data-ttu-id="ae586-129">WriteError 的管理元件</span><span class="sxs-lookup"><span data-stu-id="ae586-129">System.Management.Automation.Cmdlet.WriteError</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.WriteError)

[<span data-ttu-id="ae586-130">Windows PowerShell 錯誤記錄</span><span class="sxs-lookup"><span data-stu-id="ae586-130">Windows PowerShell Error Records</span></span>](./windows-powershell-error-records.md)

[<span data-ttu-id="ae586-131">撰寫 Windows PowerShell Cmdlet</span><span class="sxs-lookup"><span data-stu-id="ae586-131">Writing a Windows PowerShell Cmdlet</span></span>](./writing-a-windows-powershell-cmdlet.md)
